File: packages/dart/test/conformance_test.dart
// Conformance suite for the Dart (canonical) parser.
//
//  • "ground truth" group  — the 3 fixtures whose OCR was manually verified
//    against the real photos. These assert real-world correctness.
//  • "golden parity" group — all 46 fixtures must match spec/golden/expected.json,
//    the frozen output of this same canonical parser. The Python/TS ports run an
//    equivalent suite against the SAME golden, guaranteeing cross-language parity.
import 'dart:convert';
import 'dart:io';

import 'package:costco_price_tag_parser/costco_price_tag_parser.dart';
import 'package:test/test.dart';

void main() {
  final root = _repoRoot();
  final fixtures = jsonDecode(
    File('$root/spec/fixtures/ocr_raw.json').readAsStringSync(),
  ) as Map<String, dynamic>;

  PriceTagData parse(String fn) {
    final lines = (fixtures[fn] as List)
        .map((e) => OcrLine.fromJson(e as Map<String, dynamic>))
        .toList();
    return parsePriceTag(lines);
  }

  group('ground truth (human-verified photos)', () {
    test('IMG_3374 — discount tag', () {
      final t = parse('IMG_3374.JPG');
      expect(t.itemNumber, '685246');
      expect(t.plusMark, isTrue);
      expect(t.finalPrice, 349900);
      expect(t.tagType, TagType.discount);
      expect(t.discount?.originalPrice, 389900);
      expect(t.discount?.discountAmount, 40000);
    });

    test('IMG_3379 — regular tag with unit price', () {
      final t = parse('IMG_3379.JPG');
      expect(t.itemNumber, '649221');
      expect(t.finalPrice, 23890);
      expect(t.tagType, TagType.regular);
      expect(t.unitPrice?.value, 11945);
      expect(t.unitPrice?.unit, '개');
    });

    test('IMG_3383 — regular tag with unit price', () {
      final t = parse('IMG_3383.JPG');
      expect(t.itemNumber, '654718');
      expect(t.finalPrice, 119900);
      expect(t.unitPrice?.value, 11990);
      expect(t.unitPrice?.unit, '100G');
    });
  });

  group('golden parity (all fixtures)', () {
    final golden = jsonDecode(
      File('$root/spec/golden/expected.json').readAsStringSync(),
    ) as Map<String, dynamic>;

    test('golden covers every fixture', () {
      expect(golden.keys.toSet(), fixtures.keys.toSet());
    });

    for (final fn in fixtures.keys.toList()..sort()) {
      test(fn, () {
        final got = parse(fn).toJson()..remove('rawText');
        expect(_canon(got), _canon(golden[fn]),
            reason: 'parser output drifted from frozen golden for $fn');
      });
    }
  });

  // Synthetic inputs that the real fixtures don't exercise — they pin the
  // cross-language hazards the review surfaced (see spec/SPEC.md "parity notes").
  group('parity traps (synthetic inputs)', () {
    test('yTop tie resolves to the left-most item (deterministic)', () {
      // Right-most listed first; without the x tiebreak a naive sort keeps it.
      final tag = parsePriceTag(const [
        OcrLine(text: '222222', x: 0.50, yTop: 0.30, w: 0.1, h: 0.05, conf: 1),
        OcrLine(text: '111111', x: 0.10, yTop: 0.30, w: 0.1, h: 0.05, conf: 1),
      ]);
      expect(tag.itemNumber, '111111');
    });

    test('non-ASCII space (NBSP) before 원 still yields the price', () {
      final tag = parsePriceTag(const [
        OcrLine(text: '685246', x: 0.3, yTop: 0.1, w: 0.1, h: 0.03, conf: 1),
        OcrLine(text: '12345 원', x: 0.3, yTop: 0.5, w: 0.2, h: 0.10, conf: 1),
      ]);
      expect(tag.finalPrice, 12345);
      expect(tag.itemNumber, '685246');
    });
  });
}

/// Order-independent canonical JSON string for deep comparison.
String _canon(Object? v) {
  if (v is Map) {
    final keys = v.keys.map((k) => k.toString()).toList()..sort();
    return '{${keys.map((k) => '${jsonEncode(k)}:${_canon(v[k])}').join(',')}}';
  }
  if (v is List) return '[${v.map(_canon).join(',')}]';
  return jsonEncode(v);
}

String _repoRoot() {
  var dir = Directory.current;
  while (true) {
    if (Directory('${dir.path}/spec/fixtures').existsSync()) return dir.path;
    final parent = dir.parent;
    if (parent.path == dir.path) {
      throw StateError('repo root not found from ${Directory.current.path}');
    }
    dir = parent;
  }
}

File: packages/typescript/test/conformance.test.ts
/**
 * Conformance suite — same fixtures + golden as the Dart/Python ports.
 * See ../../../spec/CONFORMANCE.md.
 */

import { describe, expect, it } from "vitest";
import { existsSync, readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";
import { parsePriceTag, type OcrLine } from "../src/index.js";

function repoRoot(): string {
  let dir = dirname(fileURLToPath(import.meta.url));
  while (!existsSync(join(dir, "spec", "fixtures"))) {
    const parent = dirname(dir);
    if (parent === dir) throw new Error("repo root (with spec/fixtures) not found");
    dir = parent;
  }
  return dir;
}

const ROOT = repoRoot();
const FIXTURES: Record<string, OcrLine[]> = JSON.parse(
  readFileSync(join(ROOT, "spec/fixtures/ocr_raw.json"), "utf8"),
);
const GOLDEN: Record<string, Record<string, unknown>> = JSON.parse(
  readFileSync(join(ROOT, "spec/golden/expected.json"), "utf8"),
);

const parse = (fn: string) => parsePriceTag(FIXTURES[fn]);

describe("ground truth (human-verified photos)", () => {
  it("IMG_3374 — discount tag", () => {
    const t = parse("IMG_3374.JPG");
    expect(t.itemNumber).toBe("685246");
    expect(t.plusMark).toBe(true);
    expect(t.finalPrice).toBe(349900);
    expect(t.tagType).toBe("discount");
    expect(t.discount?.originalPrice).toBe(389900);
    expect(t.discount?.discountAmount).toBe(40000);
  });

  it("IMG_3379 — regular tag with unit price", () => {
    const t = parse("IMG_3379.JPG");
    expect(t.itemNumber).toBe("649221");
    expect(t.finalPrice).toBe(23890);
    expect(t.tagType).toBe("regular");
    expect(t.unitPrice?.value).toBe(11945);
    expect(t.unitPrice?.unit).toBe("개");
  });

  it("IMG_3383 — regular tag with unit price", () => {
    const t = parse("IMG_3383.JPG");
    expect(t.itemNumber).toBe("654718");
    expect(t.finalPrice).toBe(119900);
    expect(t.unitPrice?.value).toBe(11990);
    expect(t.unitPrice?.unit).toBe("100G");
  });
});

describe("golden parity (all fixtures)", () => {
  it("golden covers every fixture", () => {
    expect(new Set(Object.keys(GOLDEN))).toEqual(new Set(Object.keys(FIXTURES)));
  });

  for (const fn of Object.keys(FIXTURES).sort()) {
    it(fn, () => {
      const { rawText: _rawText, ...got } = parse(fn);
      expect(got).toEqual(GOLDEN[fn]);
    });
  }
});

// Synthetic inputs that the real fixtures don't exercise — they pin the
// cross-language hazards the review surfaced (see spec/SPEC.md "parity notes").
describe("parity traps (synthetic inputs)", () => {
  it("yTop tie resolves to the left-most item (deterministic)", () => {
    // Right-most listed first; without the x tiebreak a naive sort keeps it.
    const tag = parsePriceTag([
      { text: "222222", x: 0.5, yTop: 0.3, w: 0.1, h: 0.05, conf: 1 },
      { text: "111111", x: 0.1, yTop: 0.3, w: 0.1, h: 0.05, conf: 1 },
    ]);
    expect(tag.itemNumber).toBe("111111");
  });

  it("non-ASCII space (NBSP) before 원 still yields the price", () => {
    const tag = parsePriceTag([
      { text: "685246", x: 0.3, yTop: 0.1, w: 0.1, h: 0.03, conf: 1 },
      { text: "12345 원", x: 0.3, yTop: 0.5, w: 0.2, h: 0.1, conf: 1 },
    ]);
    expect(tag.finalPrice).toBe(12345);
    expect(tag.itemNumber).toBe("685246");
  });

  it("a nutrition/weight figure is not taken as the price", () => {
    // "1,584g" is taller (bigger h) — without the weight guard it would win.
    const tag = parsePriceTag([
      { text: "685246", x: 0.3, yTop: 0.1, w: 0.1, h: 0.03, conf: 1 },
      { text: "1,584g", x: 0.3, yTop: 0.4, w: 0.2, h: 0.2, conf: 1 },
      { text: "9,900원", x: 0.3, yTop: 0.6, w: 0.2, h: 0.1, conf: 1 },
    ]);
    expect(tag.finalPrice).toBe(9900);
    expect(tag.itemNumber).toBe("685246");
  });

  it("a 6-digit run inside a longer number is not an item number", () => {
    expect(
      parsePriceTag([{ text: "ITEM: 3663092", x: 0.1, yTop: 0.3, w: 0.3, h: 0.05, conf: 1 }])
        .itemNumber,
    ).toBeNull();
    // …but a properly bounded 6-digit run still is.
    expect(
      parsePriceTag([{ text: "#512905", x: 0.1, yTop: 0.3, w: 0.2, h: 0.05, conf: 1 }]).itemNumber,
    ).toBe("512905");
  });

  it("a calendar-implausible date is dropped", () => {
    const tag = parsePriceTag([
      { text: "685246", x: 0.3, yTop: 0.1, w: 0.1, h: 0.03, conf: 1 },
      { text: "할인행사", x: 0.3, yTop: 0.2, w: 0.2, h: 0.03, conf: 1 },
      { text: "10,000원", x: 0.3, yTop: 0.5, w: 0.2, h: 0.2, conf: 1 },
      { text: "20,000원", x: 0.3, yTop: 0.3, w: 0.2, h: 0.05, conf: 1 },
      { text: "2026.13.45", x: 0.3, yTop: 0.7, w: 0.2, h: 0.03, conf: 1 },
    ]);
    expect(tag.tagType).toBe("discount");
    expect(tag.discount?.originalPrice).toBe(20000);
    expect(tag.discount?.periodStart).toBeNull();
    expect(tag.discount?.periodEnd).toBeNull();
  });
});
